src/main/java/gregtech/api/pipenet/longdist/LongDistanceNetwork.java

```diff
@@ -0,0 +1,353 @@
+package gregtech.api.pipenet.longdist;
+
+import gregtech.api.pipenet.WorldPipeNet;
+import it.unimi.dsi.fastutil.objects.Object2ObjectOpenHashMap;
+import it.unimi.dsi.fastutil.objects.ObjectOpenHashSet;
+import net.minecraft.nbt.NBTBase;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.nbt.NBTTagList;
+import net.minecraft.nbt.NBTTagLong;
+import net.minecraft.util.EnumFacing;
+import net.minecraft.util.math.BlockPos;
+import net.minecraft.world.World;
+import net.minecraft.world.storage.WorldSavedData;
+import net.minecraftforge.common.util.Constants;
+
+import javax.annotation.Nonnull;
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+public class LongDistanceNetwork {
+
+    public static LongDistanceNetwork get(World world, BlockPos pos) {
+        return WorldData.get(world).getNetwork(pos);
+    }
+
+    private final LongDistancePipeType pipeType;
+    private final WorldData world;
+    // all pipes and endpoints in this net
+    public final ObjectOpenHashSet<BlockPos> longDistancePipeBlocks = new ObjectOpenHashSet<>();
+    // stores all connected endpoints, but only the first two are being used
+    private final List<MetaTileEntityLongDistanceEndpoint> endpoints = new ArrayList<>();
+    // all endpoint positions, for nbt
+    private final List<BlockPos> endpointPoss = new ArrayList<>();
+    private boolean calculating = false;
+
+    protected LongDistanceNetwork(LongDistancePipeType pipeType, WorldData world) {
+        this.pipeType = pipeType;
+        this.world = world;
+    }
+
+    /**
+     * Calculates one or more networks based on the given starting points.
+     * For this it will start a new thread to keep the main thread free.
+     */
+    protected void recalculateNetwork(Collection<BlockPos> starts) {
+        this.calculating = true;
```

Comment:
Using a simple boolean here can cause issues since we are referencing this field across multiple threads, and could cause a data race